File: packages/core/src/gate.ts
import Database from "better-sqlite3";
import {
	getRuntimeHookValue,
	setRuntimeHookValue,
	deleteRuntimeHookValue,
} from "./storage/runtime-state";

/**
 * Returns True if stored value != given value (or no stored value).
 */
export function shouldFire(
	convId: string,
	key: string,
	value: unknown,
	conn?: Database.Database,
): boolean {
	const prev = getRuntimeHookValue(convId, -1, key, conn);
	return String(prev) !== String(value);
}

/**
 * Record that this gate has fired for this value.
 */
export function markFired(
	convId: string,
	key: string,
	value: unknown,
	conn?: Database.Database,
): void {
	setRuntimeHookValue(convId, -1, key, String(value), conn);
}

/**
 * Returns True if this exact value was already recorded (same-window dedup).
 */
export function isDuplicate(
	convId: string,
	key: string,
	value: unknown,
	conn?: Database.Database,
): boolean {
	const prev = getRuntimeHookValue(convId, -1, key, conn);
	return String(prev) === String(value);
}

/**
 * Delete old record if stored value differs from new_value (cross-window re-alert).
 */
export function clearStale(
	convId: string,
	key: string,
	newValue: unknown,
	conn?: Database.Database,
): void {
	const prev = getRuntimeHookValue(convId, -1, key, conn);
	if (prev && String(prev) !== String(newValue)) {
		deleteRuntimeHookValue(convId, -1, key, conn);
	}
}

/**
 * Returns True every 5th user input.
 */
export function shouldInjectTone(userInputCount: number): boolean {
	return userInputCount % 5 === 0;
}

File: scripts/core/gate.py
from core.storage.runtime_state import get_hook_state as _get, set_hook_state as _set

def should_fire(conv_id: str, key: str, value) -> bool:
    """Returns True if stored value != given value (or no stored value)."""
    prev = _get(conv_id, -1, key)
    return str(prev) != str(value)

def mark_fired(conv_id: str, key: str, value) -> None:
    """Record that this gate has fired for this value."""
    _set(conv_id, -1, key, str(value))

def is_duplicate(conv_id: str, key: str, value) -> bool:
    """Returns True if this exact value was already recorded (same-window dedup)."""
    prev = _get(conv_id, -1, key)
    return str(prev) == str(value)

def clear_stale(conv_id: str, key: str, new_value) -> None:
    """Delete old record if stored value differs from new_value (cross-window re-alert)."""
    prev = _get(conv_id, -1, key)
    if prev and str(prev) != str(new_value):
        from core.storage.runtime_state import delete_hook_state as _del
        _del(conv_id, -1, key)
